### ISSUE TITLE
Edge to Edge: InvalidAccessError in RTCPeerConnection.prototype._transceive

### ISSUE THREAD

**Issue by zarknight** [webrtcHacks/adapter#324]
**Browsers and versions affected**
Win10 Edge

**Description**
Error occurs: InvalidAccessError in RTCPeerConnection.prototype._transceive

> edge_shim.js (Line 329): transceiver.rtpSender.send(params);

---------------------- Local SDP -----------------------
v=0
o=thisisadapterortc 8169639915646943137 2 IN IP4 127.0.0.1
s=-
t=0 0
m=audio 9 UDP/TLS/RTP/SAVPF 104 9 106 0 8 103 97 13 118 101
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=rtpmap:104 SILK/16000
a=rtcp-fb:104 x-message app send:dsh recv:dsh
a=rtpmap:9 G722/8000
a=rtcp-fb:9 x-message app send:dsh recv:dsh
a=rtpmap:106 OPUS/48000/2
a=rtcp-fb:106 x-message app send:dsh recv:dsh
a=rtpmap:0 PCMU/8000
a=rtcp-fb:0 x-message app send:dsh recv:dsh
a=rtpmap:8 PCMA/8000
a=rtcp-fb:8 x-message app send:dsh recv:dsh
a=rtpmap:103 SILK/8000
a=rtcp-fb:103 x-message app send:dsh recv:dsh
a=rtpmap:97 RED/8000
a=rtpmap:13 CN/8000
a=rtpmap:118 CN/16000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 events=0-16
a=rtcp-mux
a=ice-ufrag:+Kv5
a=ice-pwd:18FoiLbCgj1ch9gxzyadrWtn
a=setup:actpass
a=fingerprint:sha-256 84:E8:5E:82:67:7A:D4:9C:5C:BC:86:92:1E:E1:45:63:0C:9E:D8:F4:61:AB:5C:B4:4C:68:39:33:2B:13:BF:25
a=mid:87893tda5j
a=sendrecv
a=msid:3432A451-EF7E-4B5B-8728-987DB5758CB8 8534CA81-B9B0-4D3B-8D21-58A392AFB0EE
a=ssrc:1001 msid:3432A451-EF7E-4B5B-8728-987DB5758CB8 8534CA81-B9B0-4D3B-8D21-58A392AFB0EE
a=ssrc:1001 cname:7ipo2hhuse
m=video 9 UDP/TLS/RTP/SAVPF 122 123
b=AS:300
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=rtpmap:122 X-H264UC/90000
a=fmtp:122 packetization-mode=1;mst-mode=NI-TC
a=rtcp-fb:122 x-message app send:src,x-pli recv:src,x-pli
a=rtpmap:123 x-ulpfecuc/90000
a=rtcp-mux
a=ice-ufrag:FAYS
a=ice-pwd:po4j3EVXnxxLiNayfFN1v/gP
a=setup:actpass
a=fingerprint:sha-256 70:D4:F9:23:03:17:04:CF:58:87:89:2F:80:12:0C:60:39:68:7D:0A:6F:EF:9F:17:3C:BE:84:00:79:6E:FC:E4
a=mid:cllnarqpdu
a=sendrecv
a=msid:3432A451-EF7E-4B5B-8728-987DB5758CB8 8B4222C9-E612-4CFE-A80E-168666427882
a=ssrc:3003 msid:3432A451-EF7E-4B5B-8728-987DB5758CB8 8B4222C9-E612-4CFE-A80E-168666427882
a=ssrc:3003 cname:7ipo2hhuse

---------------------- Remote SDP -----------------------
v=0
o=- 0 0 IN IP4 127.0.0.1
s=TestMCU
t=0 0
m=audio 1 RTP/SAVPF 0 101
c=IN IP4 0.0.0.0
a=rtcp:1 IN IP4 0.0.0.0
a=candidate:1 1 udp 2013266431 192.168.11.100 57199 typ host generation 0
a=ice-ufrag:tY4D
a=ice-pwd:/8OdldVzgTebXkDw9oauEZ
a=fingerprint:sha-256 ED:25:B7:B2:33:6F:7B:4F:89:35:A8:02:30:C3:BD:AD:FC:17:58:DE:94:05:36:3A:70:F7:06:77:04:D1:67:FF
a=sendrecv
a=extmap:1 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=rtcp-mux
a=rtpmap:0 PCMU/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 events=0-16
a=maxptime:60
a=ssrc:44444 cname:o/i14u9pJrxRKAsu
a=ssrc:44444 msid:pAuwXq02dX a0
a=ssrc:44444 mslabel:pAuwXq02dX
a=ssrc:44444 label:pAuwXq02dXa0
m=video 1 RTP/SAVPF 
c=IN IP4 0.0.0.0
a=rtcp:1 IN IP4 0.0.0.0
a=candidate:1 1 udp 2013266431 192.168.11.100 57910 typ host generation 0
a=ice-ufrag:JJ8A
a=ice-pwd:YOLDRidiZFk4UiOLXcVpKb
a=extmap:2 urn:ietf:params:rtp-hdrext:toffset
a=extmap:3 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=fingerprint:sha-256 ED:25:B7:B2:33:6F:7B:4F:89:35:A8:02:30:C3:BD:AD:FC:17:58:DE:94:05:36:3A:70:F7:06:77:04:D1:67:FF
a=sendrecv
a=rtcp-mux
a=ssrc:55543 cname:o/i14u9pJrxRKAsu
a=ssrc:55543 msid:pAuwXq02dX v0
a=ssrc:55543 mslabel:pAuwXq02dX
a=ssrc:55543 label:pAuwXq02dXv0

**Comment by fippo** [comment]
err... you are not offering any video codecs supported by Edge in the answer. This means the common codecs list is empty which means calling send() will fail.
Your MCU should reject the m-line in that case I think even though I agree that the shim should catch this.

**Comment by aszmyd** [comment]
@zarknight were You able to finally run it after @fippo comment? I'm a beginner here and i also have this issue. @fippo what do You mean by the `codecs list is empty`? How can i setup some codecs?

**Comment by fippo** [comment]
@aszmyd in order to be able to talk to each other both peers need to have a common video codec. Edge supports H264UC (danger zone) and, in recent Windows Insider releases, H264. It does not (yet) support VP8 and VP9.

**Comment by aszmyd** [comment]
@fippo thanks for reply. Btw can You point me some way of grabbing Edge rtc internals like it is available on Chrome using `chrome://webrtc-internals/` ? We have some solution that works on Chrome and FF but it doesnt work with Edge yet and i need to debug what events are sended/received.

**Comment by Retropotenza** [comment]
I've tried to stream H264 over WebRTC on W10 1607, build 14936.1000 (insider slow ring) and W10 1607, build 160923.1700 (insider fast ring), but both list only X-H264UC and x-ulpfecuc as available video codecs. It seems that H264/AVC for RTC is still in development (https://developer.microsoft.com/en-us/microsoft-edge/platform/status/h264avcforrtc/?filter=f3f0000bf&search=rtc&q=h264). It is working in Firefox and Chrome (in Chrome for rendering H264 over WebRTC I had to disable the hardware acceleration, on both Windows and macOS).

**Comment by mondain** [comment]
MS Edge 40 / edgeHTML 15 supports everything you'd expect in an sdp offer now:
```
v=0
o=thisisadapterortc 8169639915646943137 2 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE 9p2wvtyahv t9c6mhpao7
a=ice-options:trickle
m=audio 9 UDP/TLS/RTP/SAVPF 104 102 9 0 8 103 97 13 118 101
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=rtpmap:104 SILK/16000
a=rtcp-fb:104 x-message app send:dsh recv:dsh
a=rtpmap:102 opus/48000/2
a=rtcp-fb:102 x-message app send:dsh recv:dsh
a=rtpmap:9 G722/8000
a=rtcp-fb:9 x-message app send:dsh recv:dsh
a=rtpmap:0 PCMU/8000
a=rtcp-fb:0 x-message app send:dsh recv:dsh
a=rtpmap:8 PCMA/8000
a=rtcp-fb:8 x-message app send:dsh recv:dsh
a=rtpmap:103 SILK/8000
a=rtcp-fb:103 x-message app send:dsh recv:dsh
a=rtpmap:97 RED/8000
a=rtpmap:13 CN/8000
a=rtpmap:118 CN/16000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 events=0-16
a=maxptime:100
a=rtcp-mux
a=extmap:1 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:3 http://skype.com/experiments/rtp-hdrext/fast_bandwidth_feedback#version_2
a=ice-ufrag:lMRF
a=ice-pwd:NkRjfmT4UzuYaOo/xIKa0iuT
a=setup:actpass
a=fingerprint:sha-256 09:0C:4F:FC:C6:54:B9:75:0D:CB:6D:91:9B:F2:9B:8C:BE:75:84:82:32:28:FE:DA:29:AE:50:53:1E:ED:D1:75
a=mid:9p2wvtyahv
a=sendrecv
a=msid:13E696D1-5F95-42F5-B149-887B0448FF19 F7DE95AA-B6D8-4242-882D-477CDE3660FF
a=ssrc:1001 msid:13E696D1-5F95-42F5-B149-887B0448FF19 F7DE95AA-B6D8-4242-882D-477CDE3660FF
a=ssrc:1001 cname:z1gldgcaba
a=rtcp-rsize
m=video 9 UDP/TLS/RTP/SAVPF 122 107 100 99 96 123
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=rtpmap:122 X-H264UC/90000
a=fmtp:122 packetization-mode=1;mst-mode=NI-TC
a=rtcp-fb:122 x-message app send:src,x-pli recv:src,x-pli
a=rtpmap:107 H264/90000
a=fmtp:107 profile-level-id=42C02A;packetization-mode=1;level-asymmetry-allowed=1
a=rtcp-fb:107 nack
a=rtcp-fb:107 nack pli
a=rtcp-fb:107 goog-remb
a=rtpmap:100 VP8/90000
a=rtcp-fb:100 nack
a=rtcp-fb:100 nack pli
a=rtcp-fb:100 goog-remb
a=rtpmap:99 rtx/90000
a=fmtp:99 apt=107;rtx-time=3000
a=rtpmap:96 rtx/90000
a=fmtp:96 apt=100;rtx-time=3000
a=rtpmap:123 x-ulpfecuc/90000
a=rtcp-mux
a=extmap:1 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:3 http://skype.com/experiments/rtp-hdrext/fast_bandwidth_feedback#version_2
a=ice-ufrag:xmbm
a=ice-pwd:eCQ+vl3g5OeAoK4rPUdRYxb2
a=setup:actpass
a=fingerprint:sha-256 F2:85:9C:27:E1:22:EF:93:F0:63:D8:28:B0:AD:F1:A4:83:A7:05:F2:62:D6:EF:F4:2B:7A:A4:81:63:A3:D4:0E
a=mid:t9c6mhpao7
a=sendrecv
a=msid:13E696D1-5F95-42F5-B149-887B0448FF19 46D50AC9-1B33-4895-8196-CABB2501FF28
a=ssrc:3003 msid:13E696D1-5F95-42F5-B149-887B0448FF19 46D50AC9-1B33-4895-8196-CABB2501FF28
a=ssrc:3004 msid:13E696D1-5F95-42F5-B149-887B0448FF19 46D50AC9-1B33-4895-8196-CABB2501FF28
a=ssrc-group:FID 3003 3004
a=ssrc:3003 cname:z1gldgcaba
a=ssrc:3004 cname:z1gldgcaba
a=rtcp-rsize
```
You may answer back with PCM (your example answer above) or Opus and for video VP8 or H264 (not H264UC unless you support it on your MCU)

**Comment by aszmyd** [comment]
@mondain thanks to You'r SDP example i've been able to find why Edge gives `InvalidAccessError` in our SDP. Turned out that it was just because of how we're "labeling" ssrc line:

    v=0
    o=- 5019250734634223696 1 IN IP4 0.0.0.0
    s=-
    t=0 0
    a=ice-lite
    a=group:BUNDLE video_receiving
    m=video 1 RTP/SAVPF 101
    a=rtpmap:101 VP8/90000
    c=IN IP4 52.207.193.2
    a=candidate:1 1 udp 1 52.206.120.2 6001 typ host generation 0
    a=mid:video_receiving
    a=ice-pwd:6c1bsh8fAdclnsd6nbs3f9q3g7
    a=ice-ufrag:38kt31bhcbmfa5
    a=ice-options:google-ice
    a=fingerprint:sha-256 56:DC:0C:3C:44:E8:84:2C:4A:07:E8:03:A3:A2:DC:F7:C1:D1:72:B8:5F:00:2E:4A:D2:B3:F8:82:7D:59:24:75
    a=setup:actpass
    a=rtcp-mux
    a=sendonly
    a=ssrc:2615978094 mslabel: video_receiving
    a=ssrc:2615978094 label: video_receiving
    a=ssrc:2615978094 cname: video_receiving  <-- ONLY THIS HAS BEEN ADDED

Looks like Edge uses **cname**. Without it, such SDP gives **InvalidAccessError** when trying to use it.

**Comment by fippo** [comment]
well, adapter relies on at least one cname being present to figure out the ssrc the peer is sending on. Should probably throw a more explicit error if that is missing.

**Comment by fippo** [comment]
should be fixed as of rtcpeerconnection-shim@1.2.1 (see https://github.com/otalk/rtcpeerconnection-shim/pull/40)
